### PR TITLE
In Progress - Added a linked template for referencing secret and certificate with dynamic ID in mainTemplate.json and 3 required parameters in mainTemplate.json and createUiDefinition.json

### PR DESCRIPTION
**Pull request by ajitesh-s**
# PR Checklist

Check these items before submitting a PR... 

[Contribution Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md)

[Best Practice Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md)

- [ ] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog

*
*
*

### CODE REVIEW

File: lansa-vmss-windows-autoscale-sql-database/createUiDefinition.json [hunk 1]
```diff
@@ -188,7 +209,48 @@
...
+                            "regex": "^(?!\\s*$).+",
+                            "validationMessage": "The value must not be empty."
+                        },
+                        "visible": "[equals(steps('SecurityConfig').keyVaultEnabledForTemplateDeployment, 'true')]"
+                    },
+                    {
+                        "name": "keyVaultCertificateID",
+                        "type": "Microsoft.Common.TextBox",
+                        "label": "Key Vault Certificate ID",
+                        "toolTip": "Base 64 encoded String representing the SSL certificate",
```
Comment (robe070) [inline lansa-vmss-windows-autoscale-sql-database/createUiDefinition.json]: The tooltip concerns me. This is an ID, not a certificate string

File: lansa-vmss-windows-autoscale-sql-database/createUiDefinition.json [hunk 2]
```diff
@@ -188,7 +209,48 @@
...
+                        "constraints": {
+                            "required": true,
+                            "regex": "^(?!\\s*$).+",
+                            "validationMessage": "The value must not be empty."
+                        },
+                        "visible": "[equals(steps('SecurityConfig').keyVaultEnabledForTemplateDeployment, 'true')]"
+                    },
+                    {
+                        "name": "keyVaultSecretID",
```
Comment (robe070) [inline lansa-vmss-windows-autoscale-sql-database/createUiDefinition.json]: I'm pretty sure there is no Secret for a Key Vault Certificate. You have permission to use it, or you don't.

File: lansa-vmss-windows-autoscale-sql-database/mainTemplate.json [hunk 3]
```diff
@@ -454,6 +487,50 @@
...
+            
+          }
+        },
+	"parameters": {
+          "vaultCertificate": {
+            "reference": {
+              "keyVault": {
+                "id": "[resourceId(subscription().subscriptionId, resourceGroup().name, 'Microsoft.KeyVault/vaults', parameters('vaultName'))]"
+			},
+              "certificateName": "[parameters('keyVaultCertificateID')]"
```
Comment (robe070) [inline lansa-vmss-windows-autoscale-sql-database/mainTemplate.json]: I believe from the documentation that the certificate Name is required, as coded here. But you have asked for the Certificate ID in the UI, which is incorrect.

File: lansa-vmss-windows-autoscale-sql-database/mainTemplate.json [hunk 4]
```diff
@@ -676,7 +753,7 @@
                "osProfile": {
                   "computerNamePrefix": "[variables('dbvmssName')]",
                   "adminUsername": "[parameters('adminUsername')]",
-                  "adminPassword": "[parameters('adminPassword')]",
+                  "vaultSecretID": "[parameters('vaultSecretID')]",
```
Comment (robe070) [inline lansa-vmss-windows-autoscale-sql-database/mainTemplate.json]: Please explain this. Key Vault is nothing to do with the VM

**Review by robe070 — CHANGES_REQUESTED** [review]
Great to see your progress!
Please respond to the comments. Its critical that you handle a renewed certificate automatically. 

And of course you probably already realise, but just to make sure, when you need to finalise the PR, you'll need to resolve the merge conflict.

File: lansa-vmss-windows-autoscale-sql-database/createUiDefinition.json (hunk 1 above)
Comment (ajitesh-s) [reply lansa-vmss-windows-autoscale-sql-database/createUiDefinition.json]: Will change that to Certificate Name

File: lansa-vmss-windows-autoscale-sql-database/createUiDefinition.json (hunk 2 above)
Comment (ajitesh-s) [reply lansa-vmss-windows-autoscale-sql-database/createUiDefinition.json]: Yes I agree with that, now SecretKey parameter is removed.

File: lansa-vmss-windows-autoscale-sql-database/mainTemplate.json (hunk 3 above)
Comment (ajitesh-s) [reply lansa-vmss-windows-autoscale-sql-database/mainTemplate.json]: Sure, will change that to "keyVaultCertificateName"

File: lansa-vmss-windows-autoscale-sql-database/mainTemplate.json (hunk 4 above)
Comment (ajitesh-s) [reply lansa-vmss-windows-autoscale-sql-database/mainTemplate.json]: Changed "adminPassword" with "vaultSecretId" by mistake here. I have reverted this change.

**Review by robe070 — COMMENTED** [review]
Looks good. How did you go with deploying the AG with a certificate with a password without supplying the password?

**Comment by ajitesh-s** [comment]
> Looks good. How did you go with deploying the AG with a certificate with a password without supplying the password?

Seems like we do not require a password, even for a cert with password. So existing code works fine even with the new cert you created.

**Comment by ajitesh-s** [comment]
> > Looks good. How did you go with deploying the AG with a certificate with a password without supplying the password?
> 
> Seems like we do not require a password, even for a cert with password. So existing code works fine even with the new cert you created.

Just need to remove the params from output section and changes will be complete by then

**Pull request merged**